### PR TITLE
[JENKINS-68602] parent name for ProjectNamingStrategy

### PR DESCRIPTION
**Pull request by mawinter69**
add a new method checkName(parentName, name) to the ProjectNamingStrategy so that
it is possible to check the full name of an item.

See [JENKINS-68602](https://issues.jenkins.io/browse/JENKINS-68602).

### Proposed changelog entries

* Developer: Enable ProjectNamingStrategy to check the full name of a new item

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6598"></a>

### CODE REVIEW

File: core/src/main/java/hudson/model/Job.java
Comment (alecharp) [inline core/src/main/java/hudson/model/Job.java:1354]: Just to comply with Checkstyle

```suggestion
                    namingStrategy.checkName(getParent().getFullName(), name);
```

File: core/src/main/java/jenkins/model/ProjectNamingStrategy.java
Comment (alecharp) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java:71]: Same, checkstyle is unhappy about this

```suggestion
    @Deprecated
```

File: core/src/main/java/jenkins/model/ProjectNamingStrategy.java
Comment (alecharp) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java:88]: To keep the same indentation in the file

```suggestion
    public void checkName(String parentName, String name) throws Failure {
        checkName(name);
    }
```

**Review by alecharp — CHANGES_REQUESTED** [review]
The default implementation of the new method is not using the new parameter, and the implementation in core are not using it either. Do you have a downstream pull request that is using the new parameter? I don't see the use nor need for this new API at the moment.

**Comment by basil** [comment]
The intended consumer https://github.com/jenkinsci/role-strategy-plugin/pull/179 has not yet been reviewed. That needs to happen before this core PR can be considered.

**Comment by NotMyFault** [comment]
> The intended consumer [jenkinsci/role-strategy-plugin#179](https://github.com/jenkinsci/role-strategy-plugin/pull/179) has not yet been reviewed. That needs to happen before this core PR can be considered.

The core PR may end up stalled, there are currently no active maintainers of the role strategy plugin.

**Comment by NotMyFault** [comment]
I removed the `stalled` label, the plugin has been adopted by the PR author.

**Comment by NotMyFault** [comment]
The downstream PR has been merged by now.

**Review by timja — COMMENTED** [review]
can you create a PR on role strategy consuming this API please?

I see you've merged a workaround, but generally we require a PR demonstrating use of the API consuming the core incremental.

apart from that this looks good to me

**Comment by mawinter69** [comment]
Role strategy is already implementing the API
Just the @Override annotation is not there yet
https://github.com/jenkinsci/role-strategy-plugin/blob/master/src/main/java/org/jenkinsci/plugins/rolestrategy/RoleBasedProjectNamingStrategy.java#L72
The workaround mentioned is for the a case when the old method is called.

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
Thanks!

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
